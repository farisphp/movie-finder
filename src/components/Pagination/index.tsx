import { Link } from 'react-router-dom'
import classnames from 'classnames'
import { generatePageUrl, generatePagination } from 'utils'

interface IPagination {
  url: string
  totalItems: number
  page: number
  size: number
}

function Pagination({ url, totalItems = 0, page, size }: IPagination) {
  const maxPage = Math.ceil(totalItems / size)
  return (
    <div className="flex justify-center space-x-6">
      <Link
        to={generatePageUrl(
          window.location.toString(),
          page - 1 < 1 ? 1 : page - 1
        )}
        className={classnames(' text-blue-400 hover:underline', {
          hidden: page === 1
        })}
      >
        Prev
      </Link>
      <div className="flex justify-center space-x-2">
        {generatePagination(url, totalItems, page, size).map(
          ({ page: pageNumber, pageUrl }) => (
            <Link key={`page-link-${pageNumber}`} to={pageUrl}>
              <span
                className={`${
                  pageNumber !== page ? 'text-blue-400 hover:underline' : ''
                }`}
              >
                {pageNumber}
              </span>
            </Link>
          )
        )}
      </div>

      <Link
        to={generatePageUrl(
          window.location.toString(),
          page + 1 > maxPage ? maxPage : page + 1
        )}
        className={classnames('text-blue-400 hover:underline', {
          hidden: maxPage === page
        })}
      >
        Next
      </Link>
    </div>
  )
}

export default Pagination
