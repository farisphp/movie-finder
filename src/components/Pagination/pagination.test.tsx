import { render, screen } from '@testing-library/react'
import { BrowserRouter } from 'react-router-dom'

import Pagination from '.'

describe('<Pagination />', () => {
  it('should render the pagination with 1 page', () => {
    const { container } = render(
      <Pagination
        url="http://127.0.0.1:5173/search?q=naruto&p=1"
        totalItems={10}
        page={1}
        size={20}
      />,
      { wrapper: BrowserRouter }
    )

    expect(screen.getByRole('link', { name: '1' })).toBeInTheDocument()
    expect(screen.queryByRole('link', { name: '2' })).not.toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render the pagination with 2 page', () => {
    const { container } = render(
      <Pagination
        url="http://127.0.0.1:5173/search?q=naruto&p=1"
        totalItems={30}
        page={1}
        size={20}
      />,
      { wrapper: BrowserRouter }
    )

    expect(screen.getByRole('link', { name: '1' })).toBeInTheDocument()
    expect(screen.getByRole('link', { name: '2' })).toBeInTheDocument()
    expect(screen.queryByRole('link', { name: '3' })).not.toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render the pagination with 10 page', () => {
    const { container } = render(
      <Pagination
        url="http://127.0.0.1:5173/search?q=naruto&p=1"
        totalItems={1000}
        page={1}
        size={20}
      />,
      { wrapper: BrowserRouter }
    )

    expect(screen.getByRole('link', { name: '1' })).toBeInTheDocument()
    expect(screen.getByRole('link', { name: '10' })).toBeInTheDocument()
    expect(screen.queryByRole('link', { name: '11' })).not.toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render the pagination with previous link', () => {
    const { container } = render(
      <Pagination
        url="http://127.0.0.1:5173/search?q=naruto&p=1"
        totalItems={50}
        page={2}
        size={20}
      />,
      { wrapper: BrowserRouter }
    )

    expect(screen.getByRole('link', { name: 'Prev' })).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render the pagination with next link', () => {
    const { container } = render(
      <Pagination
        url="http://127.0.0.1:5173/search?q=naruto&p=1"
        totalItems={30}
        page={1}
        size={20}
      />,
      { wrapper: BrowserRouter }
    )

    expect(screen.getByRole('link', { name: 'Next' })).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })
})
