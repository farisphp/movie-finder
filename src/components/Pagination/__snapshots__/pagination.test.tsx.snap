// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Pagination /> > should render the pagination with 1 page 1`] = `
<div
  class="flex justify-center space-x-6"
>
  <a
    class=" text-blue-400 hover:underline hidden"
    href="blank?about%3Ablank=&p=1"
  >
    Prev
  </a>
  <div
    class="flex justify-center space-x-2"
  >
    <a
      href="http://127.0.0.1:5173/search?q=naruto&p=1"
    >
      <span
        class=""
      >
        1
      </span>
    </a>
  </div>
  <a
    class="text-blue-400 hover:underline hidden"
    href="blank?about%3Ablank=&p=0.5"
  >
    Next
  </a>
</div>
`;

exports[`<Pagination /> > should render the pagination with 2 page 1`] = `
<div
  class="flex justify-center space-x-6"
>
  <a
    class=" text-blue-400 hover:underline hidden"
    href="blank?about%3Ablank=&p=1"
  >
    Prev
  </a>
  <div
    class="flex justify-center space-x-2"
  >
    <a
      href="http://127.0.0.1:5173/search?q=naruto&p=1"
    >
      <span
        class=""
      >
        1
      </span>
    </a>
    <a
      href="http://127.0.0.1:5173/search?q=naruto&p=2"
    >
      <span
        class="text-blue-400 hover:underline"
      >
        2
      </span>
    </a>
  </div>
  <a
    class="text-blue-400 hover:underline"
    href="blank?about%3Ablank=&p=1.5"
  >
    Next
  </a>
</div>
`;

exports[`<Pagination /> > should render the pagination with 10 page 1`] = `
<div
  class="flex justify-center space-x-6"
>
  <a
    class=" text-blue-400 hover:underline hidden"
    href="blank?about%3Ablank=&p=1"
  >
    Prev
  </a>
  <div
    class="flex justify-center space-x-2"
  >
    <a
      href="http://127.0.0.1:5173/search?q=naruto&p=1"
    >
      <span
        class=""
      >
        1
      </span>
    </a>
    <a
      href="http://127.0.0.1:5173/search?q=naruto&p=2"
    >
      <span
        class="text-blue-400 hover:underline"
      >
        2
      </span>
    </a>
    <a
      href="http://127.0.0.1:5173/search?q=naruto&p=3"
    >
      <span
        class="text-blue-400 hover:underline"
      >
        3
      </span>
    </a>
    <a
      href="http://127.0.0.1:5173/search?q=naruto&p=4"
    >
      <span
        class="text-blue-400 hover:underline"
      >
        4
      </span>
    </a>
    <a
      href="http://127.0.0.1:5173/search?q=naruto&p=5"
    >
      <span
        class="text-blue-400 hover:underline"
      >
        5
      </span>
    </a>
    <a
      href="http://127.0.0.1:5173/search?q=naruto&p=6"
    >
      <span
        class="text-blue-400 hover:underline"
      >
        6
      </span>
    </a>
    <a
      href="http://127.0.0.1:5173/search?q=naruto&p=7"
    >
      <span
        class="text-blue-400 hover:underline"
      >
        7
      </span>
    </a>
    <a
      href="http://127.0.0.1:5173/search?q=naruto&p=8"
    >
      <span
        class="text-blue-400 hover:underline"
      >
        8
      </span>
    </a>
    <a
      href="http://127.0.0.1:5173/search?q=naruto&p=9"
    >
      <span
        class="text-blue-400 hover:underline"
      >
        9
      </span>
    </a>
    <a
      href="http://127.0.0.1:5173/search?q=naruto&p=10"
    >
      <span
        class="text-blue-400 hover:underline"
      >
        10
      </span>
    </a>
  </div>
  <a
    class="text-blue-400 hover:underline"
    href="blank?about%3Ablank=&p=2"
  >
    Next
  </a>
</div>
`;

exports[`<Pagination /> > should render the pagination with next link 1`] = `
<div
  class="flex justify-center space-x-6"
>
  <a
    class=" text-blue-400 hover:underline hidden"
    href="blank?about%3Ablank=&p=1"
  >
    Prev
  </a>
  <div
    class="flex justify-center space-x-2"
  >
    <a
      href="http://127.0.0.1:5173/search?q=naruto&p=1"
    >
      <span
        class=""
      >
        1
      </span>
    </a>
    <a
      href="http://127.0.0.1:5173/search?q=naruto&p=2"
    >
      <span
        class="text-blue-400 hover:underline"
      >
        2
      </span>
    </a>
  </div>
  <a
    class="text-blue-400 hover:underline"
    href="blank?about%3Ablank=&p=1.5"
  >
    Next
  </a>
</div>
`;

exports[`<Pagination /> > should render the pagination with previous link 1`] = `
<div
  class="flex justify-center space-x-6"
>
  <a
    class=" text-blue-400 hover:underline"
    href="blank?about%3Ablank=&p=1"
  >
    Prev
  </a>
  <div
    class="flex justify-center space-x-2"
  >
    <a
      href="http://127.0.0.1:5173/search?q=naruto&p=1"
    >
      <span
        class="text-blue-400 hover:underline"
      >
        1
      </span>
    </a>
    <a
      href="http://127.0.0.1:5173/search?q=naruto&p=2"
    >
      <span
        class=""
      >
        2
      </span>
    </a>
    <a
      href="http://127.0.0.1:5173/search?q=naruto&p=3"
    >
      <span
        class="text-blue-400 hover:underline"
      >
        3
      </span>
    </a>
  </div>
  <a
    class="text-blue-400 hover:underline"
    href="blank?about%3Ablank=&p=2.5"
  >
    Next
  </a>
</div>
`;
